Add wide character support to `Text::truncate()`.

chinpei215 opened the pull request:
/cc @burzum

Text::truncate() is not useful for Japanese, Chinese, etc:

``` php
echo Text::truncate('0123456789', 9, ['ellipsis' => '…']);
echo Text::truncate('〇一二三四五六七八九', 9, ['ellipsis' => '…']);
// 01234567…
// 〇一二三四五六七…
```

It will be looking good if wide characters use two length:

``` php
echo Text::truncate('0123456789', 9, ['ellipsis' => '…', 'trimWidth' => true]);
echo Text::truncate('〇一二三四五六七八九', 9, ['ellipsis' => '…', 'trimWidth' => true]);
// 01234567…
// 〇一二三…
```

Of course, this method also supports `html` option:

``` php
echo Text::truncate('<p>〇<strong>一</strong>二三四五六七八九</p>', 9, [
    'ellipsis' => '…',
    'trimWidth' => true,
    'html' => true
]);
// <p>〇<strong>一</strong>二三…</p>
```

Note that I have used `mb_strwidth`  and `mb_strimwidth` in this implementation. They doesn't support [ambiguous characters](http://unicode.org/reports/tr11/). So this method might not give your  expected results, depending on your environment. For example, if you are using MS Gothic, `…` looks like a wide character, but `mb_strwidth('…')` would return `1`. Therefore, the above code shows `〇一二三…` (2 + 2 + 2 + 2 + 1 = 9).

chinpei215 committed:
Fix wrong argument order for assertSame

chinpei215 committed:
Add wide character support to `Text::truncate()`.

burzum commented:
@chinpei215 thank your for looking into this issue!

Well, the marketing and design guys here want to truncate the text. I think they would be OK if it is truncated at all and doesn't look totally strange. They already learned that languages are troublesome and will accept some constraints.

chinpei215 commented on tests/TestCase/Utility/TextTest.php:
```diff
@@ -535,48 +535,48 @@ public function testTruncate()
             'html' => true
         ]);
-        $expected = '<p><span style="font-size: medium;"><a>Iamates...</a></span></p>';
+        $expected = '<p><span style="font-size: medium;"><a>Iamatestwi...</a></span></p>';
```
This change is done for consistency. At the moment, Text::truncate doesn't behave consistently:
```php
echo Text::truncate('Hello', 3, ['exact' => false, 'ellipsis' => '...']);
// Hel...

echo Text::truncate('Hello', 3, ['exact' => false, 'ellipsis' => '...', 'html' => true]);
// ...
```

Refs #6259

chinpei215 commented:
> the marketing and design guys here want to truncate the text

If it is only an issue of the website, and you don't need to support legacy browsers, you might be able to use `text-overflow: ellipsis` in the CSS. It also works fine for Japanese language.

dereuromark commented:
Hehe, awesome how you had to almost recode PHP internals :P

markstory commented:
Nice work :clap:

chinpei215 committed:
Add a test for edge case. Remove an unneeded test.

In the current master branch, the following code will break the HTML
structure in the text, and will display `1...</b>` unexpectedly:
```
Text::truncate('1 <b>2 345</b>', 6, [
    'exact' => false,
    'html' => true,
    'ellipsis' => '...'
])
```
Now it works fine.

chinpei215 commented:
@dereuromark At first, I was thinking that it would be a piece of cake for me. But actually, it was difficult to support `html` option :persevere:

netstyler commented:
Great Job :thumbsup:

lorenzo commented on src/Utility/Text.php:
```diff
@@ -564,124 +565,241 @@ public static function tail($text, $length = 100, array $options = [])
     {
         $default = [
-            'ellipsis' => '...', 'exact' => true, 'html' => false
+            'ellipsis' => '...', 'exact' => true, 'html' => false, 'trimWidth' => false,
```
Can we maybe have another method called `Text::truncateWidth()` that calls this method with this option enabled? That way we can communicate better that this is possible

dereuromark replied:
I am not sure if that helps to understand why this is needed.

chinpei215 replied:
It would be worth to add `Text::truncateWidth()`, because Japanese or other East Asian users would need to set `trimWidth` to `true` always. I'll add it tomorrow.

markstory replied:
Named methods are easier for humans to work with :smile:

dereuromark replied:
Only if they are named non-confusing :)

chinpei215 committed:
Add Text::truncateWidth() method

Add more tests

dereuromark commented on src/Utility/Text.php:
```diff
@@ -649,6 +649,20 @@ public static function truncate($text, $length = 100, array $options = [])
+     * @return string Trimmed string.
+     * @see \Cake\Utility\Text::truncate()
+     */
+    public static function truncateWidth($text, $length = 100, array $options = [])
```
Not sure about the name then.
This sounds like the width of the truncate.
Isn't "truncateWihWidth" or "truncateUsingWidth" less confusing?

chinpei215 replied:
What about `widthTruncate()`? If it also sounds strange, I am willing to rename the method to `truncateWithWidth`.

lorenzo replied:
truncateByWidth?

markstory replied:
visualTruncate?

chinpei215 replied:
According to google search results, widthTruncate, truncateByWIdth and visualTruncate are used actually. Which would you prefer?

markstory replied:
I think they are all fitting. `widthTruncate` reads awkwardly in english to me. `truncateByWidth` or `visualTruncate` are :+1: by me.

dereuromark replied:
I would go with truncateByWidth then as this groups them together alphabetically.

chinpei215 replied:
Thanks. I'll rename it to `truncateByWidth`.

chinpei215 committed:
Rename truncateWidth to truncateByWidth

markstory commented:
Looks good to me.

dereuromark commented:
:+1:
